daemon/algod/api/server/v2/test/handlers_test.go

```diff
@@ -693,21 +696,374 @@ func TestPostTransaction(t *testing.T) {
+func simulateTransactionTest(t *testing.T, txnToUse int, format string, expectedCode int, enableTransactionSimulator bool) {
 	handler, c, rec, releasefunc := prepareTransactionTest(t, txnToUse, expectedCode, enableTransactionSimulator)
 	defer releasefunc()
-	err := handler.SimulateTransaction(c)
+	err := handler.SimulateTransaction(c, model.SimulateTransactionParams{Format: (*model.SimulateTransactionParamsFormat)(&format)})
 	require.NoError(t, err)
 	require.Equal(t, expectedCode, rec.Code)
 }
 
+func TestPostSimulateTransaction(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	testCases := []struct {
+		txnIndex                   int
+		format                     string
+		expectedStatus             int
+		enableTransactionSimulator bool
+	}{
+		{
+			txnIndex:                   -1,
+			format:                     "json",
+			expectedStatus:             400,
+			enableTransactionSimulator: true,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "json",
+			expectedStatus:             404,
+			enableTransactionSimulator: false,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "msgpack",
+			expectedStatus:             404,
+			enableTransactionSimulator: false,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "bad format",
+			expectedStatus:             404,
+			enableTransactionSimulator: false,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "json",
+			expectedStatus:             200,
+			enableTransactionSimulator: true,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "msgpack",
+			expectedStatus:             200,
+			enableTransactionSimulator: true,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "bad format",
+			expectedStatus:             400,
+			enableTransactionSimulator: true,
+		},
+	}
+
+	for i, testCase := range testCases {
+		testCase := testCase
+		t.Run(fmt.Sprintf("i=%d", i), func(t *testing.T) {
+			t.Parallel()
+			simulateTransactionTest(t, testCase.txnIndex, testCase.format, testCase.expectedStatus, testCase.enableTransactionSimulator)
+		})
+	}
+}
+
+func copyInnerTxnGroupIDs(t *testing.T, dst, src *model.PendingTransactionResponse) {
+	t.Helper()
+
+	// msgpack decodes to map[interface{}]interface{} while JSON decodes to map[string]interface{}
+	txn := dst.Txn["txn"]
+	switch dstTxnMap := txn.(type) {
+	case map[string]interface{}:
+		srcTxnMap := src.Txn["txn"].(map[string]interface{})
+		groupID, hasGroupID := srcTxnMap["grp"]
+		if hasGroupID {
+			dstTxnMap["grp"] = groupID
+		}
+	case map[interface{}]interface{}:
+		srcTxnMap := src.Txn["txn"].(map[interface{}]interface{})
+		groupID, hasGroupID := srcTxnMap["grp"]
+		if hasGroupID {
+			dstTxnMap["grp"] = groupID
+		}
+	}
+
+	if dst.InnerTxns == nil || src.InnerTxns == nil {
+		return
```

Comment:
no assert(dst.InnerTxns == nil && aex.InnerTxns == nil)?

Reply:
That's not necessary here. If there's a discrepancy, it will show in line 847 when we compare the expected and actual responses.